### ISSUE TITLE
noise labels

### ISSUE THREAD

**Issue by lunixbochs** [facebookresearch/wav2letter#532]
I've been trying to train a model that's more resistant to false positives caused by background noise. Right now I'm working with conv_glu/ASG. I have a bunch of short noise clips that I've given an N transcript for noise and ASG cannot figure out what do with it (literally none of them are trained). I can see why it would not like this, so I would prefer to just label all of the frames in the clip as noise.

I thought about modifying Train.cpp and detecting clips that had a noise label, and using cross entropy loss instead of ASG for them, but I realized that won't work with batching (if a noise clip is trained in the same batch as a non-noise clip), and I don't want to train on batchsize=1.

So right now I'm thinking of modifying ASG.

Any pointers on doing this with the wav2letter codebase and ASG? What about with seq2seq/ctc Transformer models?

I'm already running two(!) voice activity detectors before it gets to wav2letter, but the false positives are still happening, so I'd like the model in all of its much larger, slower, wisdom than the vads learn to not recognize dogs barking and stuff as actual words.

---

The other idea I just had, is to add a separate "noise" train/valid set flags, that train in their own epoch with binary cross entropy instead of using the criterion, but I have no idea how that would work with Transformer models

---

I did just have some minor success by changing the noise label from just `N` to `N|N|N|N` in the lexicon, then training 20 epochs of _only_ noise. Waiting for a hybrid noise/speech dataset to converge out of it before I can consider this a proper success however.

**Comment by xuqiantong** [comment]
@lunixbochs Actually, it is a good idea of using `N` as a new token, a label of noise. This way, however, we also distinguishes noise (`N`) from silence (`|`).  So, the better the transcription you provide, the easier an ASG learns the alignment, especially in the first epochs.  For example, if there is noise in the middle, it's better to have the transcription `i | like | N | apple` than `i | like | apple` or `I | like N apple`. 

This method is also what we used in training Switchboard models. We are planning to release the recipes.

**Comment by lunixbochs** [comment]
Part of what I want to train on is files that have noise by itself, without any speech. I want to show the network "true negatives" of speech, but I'm having a hard time getting w2l to accept it
